bundles/org.eclipse.equinox.common/src/org/eclipse/core/text/StringMatcher.java

```diff
@@ -141,14 +169,75 @@ public String toString() {
 	 * @throws IllegalArgumentException if {@code pattern == null}
 	 */
 	public StringMatcher(String pattern, boolean ignoreCase, boolean ignoreWildCards) {
+		this(pattern, ignoreCase, ignoreWildCards, true);
+	}
+
+	/**
+	 * StringMatcher constructor takes in a String object that is a simple pattern.
+	 * The pattern may contain '*' for 0 and many characters and '?' for exactly one
+	 * character.
+	 * <p>
+	 * Literal '*' and '?' characters must be escaped in the pattern e.g., "\*"
+	 * means literal "*", etc.
+	 * </p>
+	 * <p>
+	 * The escape character '\' is an escape only if followed by '*', '?', or '\'.
+	 * All other occurrences are taken literally.
+	 * </p>
+	 * <p>
+	 * If invoking the StringMatcher with string literals in Java, don't forget
+	 * escape characters are represented by "\\".
+	 * </p>
+	 * <p>
+	 * If {@code ignoreWords} is true, this {@code StringMatcher} matches a pattern
+	 * that may contain the wildcards '?' or '*' against a text. However, the
+	 * matching is not only done on the full text, but also on individual words from
+	 * the text, and if the pattern contains whitespace, the pattern is split into
+	 * sub-patterns and those are matched, too.
+	 * </p>
+	 * <p>
+	 * The precise rules are:
+	 * </p>
+	 * <ul>
+	 * <li>Leading and trailing whitespace in the pattern is ignored.</li>
+	 * <li>If the full pattern matches the full text, the match succeeds.</li>
+	 * <li>If the full pattern matches a single word of the text, the match
+	 * succeeds.</li>
+	 * <li>If all sub-patterns match a prefix of the whole text or any prefix of any
+	 * word, the match succeeds.</li>
+	 * <li>Otherwise, the match fails.</li>
+	 * </ul>
+	 * <p>
+	 * An empty pattern matches only an empty text, unless {@link #usePrefixMatch()}
+	 * is used, in which case it always matches.
+	 * </p>
+	 *
+	 * @param pattern         the pattern to match text against, must not be
+	 *                        {@code null}
+	 * @param ignoreCase      if true, case is ignored
+	 * @param ignoreWildCards if true, wild cards and their escape sequences are
+	 *                        ignored (everything is taken literally).
+	 * @param ignoreWords     if true, only matches against the whole text but not
+	 *                        individual words
+	 * @throws IllegalArgumentException if {@code pattern == null}
+	 * @since 3.20
+	 */
+	public StringMatcher(String pattern, boolean ignoreCase, boolean ignoreWildCards, boolean ignoreWords) {
 		if (pattern == null) {
 			throw new IllegalArgumentException();
 		}
 		fIgnoreCase = ignoreCase;
 		fIgnoreWildCards = ignoreWildCards;
-		fPattern = pattern;
-		fLength = pattern.length();
-
+		fIgnoreWords = ignoreWords;
+		if (fIgnoreWords) {
```

Comment:
With only my API hat on I think it would make more sense if we didn't add a new constructor to take an additional `boolean` for word matching.  Instead could we add a new method `public boolean matchWords(List<String> words)` or `public boolean matchWords(String[] words)`

I'm unsure how much that complicates the implementation.  I would imagine you would want to use an `AtomicReference<fParts[]>` to create the parts on demand if/when `matchWords` is called.

Reply:
>  Instead could we add a new method public boolean matchWords(List<String> words) or public boolean matchWords(String[] words)

I suppose that would work.  Though I think the method has to accept the whole text, given that the sub-patterns are matched against the text as well as the individual words. I can't really predict the impact if I were to change this behavior.

>  I would imagine you would want to use an AtomicReference<fParts[]> to create the parts on demand if/when matchWords is called.

I'm not sure dead-set on the `final` modifier, so an `AtomicReference` sounds like overkill if I can do a lazy initialization by checking whether `fParts == null`.

> I'm unsure how much that complicates the implementation.

It'll primarly make the check more verbose. So instead of `matcher.match(text)` it'll become `matcher.match(text) || matcher.matchWords(text)`. There are 4-5 Platform classes that need to be updated, so I think that's a doable amount of work.

It obviously also means that the trimming of the pattern has to be done explicitly, but that's also fine.

Reply:
> It'll primarly make the check more verbose. So instead of matcher.match(text) it'll become matcher.match(text) || matcher.matchWords(text). There are 4-5 Platform classes that need to be updated, so I think that's a doable amount of work.

Alternatively, I can also simply call `match(text)` inside `matchWords(text)` which I think is preferable.

Reply:
> > It'll primarly make the check more verbose. So instead of matcher.match(text) it'll become matcher.match(text) || matcher.matchWords(text). There are 4-5 Platform classes that need to be updated, so I think that's a doable amount of work.
> 
> Alternatively, I can also simply call `match(text)` inside `matchWords(text)` which I think is preferable.

That was my intention, make the new method do what you wanted to do with the existing method.